generate.go

```diff
@@ -0,0 +1,203 @@
+package cogs
+
+import (
+	"fmt"
+
+	"github.com/pelletier/go-toml"
+)
+
+type Cog struct {
+	Name string
+	// generate multiple envs?
+	env Env
+}
+
+type Env struct {
+	Config map[string]Cfg
+}
+
+type Cfg struct {
+	// Defaults to key name unless explicitly declared
+	Name  string
+	Value string
+	Path  string
+	// default should be Cfg key name
+	SubPath   string
+	encrypted bool
+}
+
+func (c Cfg) GenerateValue() string {
+	// if Path is empty or Value is non empty
+	if c.Path == "" || c.Value != "" {
+		return c.Value
+	}
+
+	if c.encrypted {
+		// TODO COGS-1657
+		// decrypt.Open(c.Path, c.SubPath)
+		return "|enc| " + c.Path
+	}
+	// TODO COGS-1659
+	// cogs.Open(c.Path, c.SubPath)
+	return "path." + c.Path + "/" + c.SubPath
+
+}
+
+func (c Cfg) String() string {
+	return fmt.Sprintf(`Cfg{
+	Name: %s
+	Value: %s
+	Path: %s
+	SubPath: %s
+	encrypted: %t
+}`, c.Name, c.Value, c.Path, c.SubPath, c.encrypted)
+}
+
+func (e Env) GenerateMap() map[string]string {
+	cfgMap := make(map[string]string)
+	for k, cfg := range e.Config {
+		cfgMap[k] = cfg.GenerateValue()
+	}
+	return cfgMap
+
+}
+
+// Mapper is an iterface that defines a struct able to generate a flat associative array
+type Mapper interface {
+	GenerateMap() map[string]string
+}
+
+// Generate is a top level command that takes an env argument and cogfilepath to return a string map
+func Generate(env, cogFile string) (map[string]string, error) {
+
+	tree, err := toml.LoadFile(cogFile)
+	if err != nil {
+		return nil, err
+	}
+	return generate(env, tree)
+
+}
+
+func generate(env string, tree *toml.Tree) (map[string]string, error) {
+	var cog Cog
+	var ok bool
+	var err error
+
+	// grab manifest name
+	cog.Name, ok = tree.Get("name").(string)
+	if !ok || cog.Name == "" {
+		return nil, fmt.Errorf("manifest.name string value must be present as a string")
+	}
+
+	var rawManifest map[string]interface{}
+	if err = tree.Unmarshal(&rawManifest); err != nil {
+		return nil, err
+	}
+
+	rawEnv, ok := rawManifest[env]
+	if !ok {
+		return nil, fmt.Errorf("%s environment missing from cog file", env)
+	}
+
+	cog.env.Config, err = ParseEnv(rawEnv)
+	if err != nil {
+		return nil, fmt.Errorf("%s: %s", env, err)
+	}
+
+	return cog.env.GenerateMap(), nil
+}
+
+// ParseEnv traverses an map interface to return a Cfg string map
+func ParseEnv(rawMap interface{}) (cfgMap map[string]Cfg, err error) {
```

Comment:
seeing empty interface as function parameters usually raises an eyebrow. From seeing how its used, and called, to me it would make more sense to make this a `map[string]interface{}` instead. Sound's like it's the caller's responsibility to make sure the right type was passed in, instead of this function worrying about it.

The type casting has to be done regardless, its more a whose's concern is it. What you do think?

Reply:
That's sensible, I'll do the type casting upstream.